okhttp/src/main/java/okhttp3/internal/connection/StreamAllocation.java

```diff
@@ -202,22 +205,34 @@ private RealConnection findConnection(int connectTimeout, int readTimeout, int w
           Route route = routes.get(i);
           Internal.instance.get(connectionPool, address, this, route);
           if (connection != null) {
+            foundPooledConnection = true;
+            result = connection;
             this.route = route;
-            return connection;
+            break;
           }
         }
       }
 
-      if (selectedRoute == null) {
-        selectedRoute = routeSelection.next();
+      if (!foundPooledConnection) {
+        if (selectedRoute == null) {
+          selectedRoute = routeSelection.next();
+        }
+
+        // Create a connection and assign it to this allocation immediately. This makes it possible
+        // for an asynchronous cancel() to interrupt the handshake we're about to do.
+        route = selectedRoute;
+        refusedStreamCount = 0;
+        result = new RealConnection(connectionPool, selectedRoute);
+        acquire(result);
       }
+    }
 
-      // Create a connection and assign it to this allocation immediately. This makes it possible
-      // for an asynchronous cancel() to interrupt the handshake we're about to do.
-      route = selectedRoute;
-      refusedStreamCount = 0;
-      result = new RealConnection(connectionPool, selectedRoute);
-      acquire(result);
+    // We have a connection. Either a connected one from the pool, or one we need to connect.
+    eventListener.connectionAcquired(call, result);
```

Comment:
Just noting in the case of a `Connection` we need to connect, the socket on the connection will be null.